### PR TITLE
[WFLY-9300] Add test for ejb security propagation using PLAIN

### PR DESCRIPTION
**Pull request by rpelisse**
Issue: https://issues.jboss.org/browse/JBEAP-13004
Upstream issue: https://issues.jboss.org/browse/WFLY-9300

_Note: I've just rebased the original proposal by @bmaxwell, run the full testsuite and found no issue with the changes. If there any issue with the actual implementation, I'll take a look but do mention Brad so he gets notified about it._

### CODE REVIEW

**Comment by kabir** [comment]
@rpelisse adding 'fixme', the tests fail

File: testsuite/integration/multinode/pom.xml [hunk 1]
```diff
@@ -36,6 +36,14 @@
         <ts.elytron.cli>multinode-setup.cli</ts.elytron.cli>
     </properties>
 
+    <dependencies>
+        <dependency>
+            <groupId>org.wildfly</groupId>
+            <artifactId>wildfly-testsuite-shared</artifactId>
+            <scope>test</scope>
```
Comment (rsvoboda) [inline testsuite/integration/multinode/pom.xml]: add `<version>${project.version}</version>`

**Review by rsvoboda — CHANGES_REQUESTED** [review]

File: testsuite/integration/multinode/pom.xml (hunk 1 above)
Comment (rpelisse) [reply testsuite/integration/multinode/pom.xml]: Good tip, thanks ! :)

Comment (rpelisse) [reply testsuite/integration/multinode/pom.xml]: No worries! I'll take a tip against no tip any day :) 

And I also should have say that I was trying to run without the dependency.

Comment (rpelisse) [reply testsuite/integration/multinode/pom.xml]: Hum, I ran into some weird (apparently unrelated) test failures locally. I've still pushed my changes as I want to see if those issues arises in CI. I'll check the result tomorrow.

**Comment by rpelisse** [comment]
@kabir FYI, I had to @Ignore one of the test that causes some failure. I will look into this, see if I can make it work properly, but I don't to delay the all changes set being merged just for that. If and when this will get merged, I'll create a new issue (and assign it) to me, to keep track of this disabled test (unless I fixed before then, of course).

File: testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/SecurityPropagationPlainTestCase.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+// This only works if you actually have deployments
+@ServerSetup(PlainSecurityPropagationTestCaseSetup.class)
+// This test appears to NOT clean up (or setup) properly and thus fails, plus causing other test to fail
+// Thus it is ignored for now
+@Ignore
+public class SecurityPropagationPlainTestCase extends AbstractSecurityPropagationTestCase {
```
Comment (rsvoboda) [inline testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/SecurityPropagationPlainTestCase.java]: All tests are ignored, not just one

File: testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/config/PlainServerConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+        @Override
+        public void apply(ModelControllerClient client, CLIWrapper cli) throws Exception {
+            cli.sendLine("/subsystem=elytron/sasl-authentication-factory=application-sasl-authentication:remove()");
+
+            StringBuilder sb = new StringBuilder(
+                    "/subsystem=elytron/sasl-authentication-factory=application-sasl-authentication:add(");
+            sb.append("sasl-server-factory=configured, security-domain=ApplicationDomain, ");
+            sb.append("mechanism-configurations=[");
+            // sb.append("{mechanism-name=JBOSS-LOCAL-USER, realm-mapper=local},");
```
Comment (rsvoboda) [inline testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/config/PlainServerConfiguration.java]: leftover ?

File: testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/config/Deployments.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        sb.append("</web-app>\n");
+        return new StringAsset(sb.toString());
+    }
+
+    private static StringAsset getEjbClientXml(boolean excludeLocalReceivers) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("<jboss-ejb-client xmlns=\"urn:jboss:ejb-client:1.2\">");
+        sb.append("<client-context>");
+        sb.append("  <ejb-receivers exclude-local-receiver=\"true\">");
+        // sb.append(" <ejb-receivers exclude-local-receiver=\"false\">");
```
Comment (rsvoboda) [inline testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/config/Deployments.java]: leftover ?  does it matter if exclude-local-receiver is true or false ?

**Comment by rsvoboda** [comment]
@mchoma please look into this too ... well once the PR is updated and tests are enabled

**Comment by mchoma** [comment]
Could there be some high level description of what is beiing tested by this. I am not sure if it is not duplication to tests based on AbstractSecurityContextPropagationTestBase, basically testing [2]

[2] https://access.redhat.com/documentation/en-us/red_hat_jboss_enterprise_application_platform/7.1/html/how_to_configure_server_security/securing_users_of_the_server_and_its_management_interfaces#identity_propagation_and_forwarding_with_elytron

**Comment by rpelisse** [comment]
@mchoma I'm not the author of the original tests, but I don't think duplication is an issue here. If the tests are duplicated from Elytron, they at least allow to prove that the integration with Elytron works. Also, if Wildfly changes provider (later on the future, obviously) the test will be there to ensure that the behavior is still as expected. (That being I'm said I'm no expert of the topic here).

For the record, I'm working on making the test passes. Once I've fixed it, I'll update th PR accordingly.

**Comment by mchoma** [comment]
@bmaxwell, please see my question ^^

AbstractSecurityContextPropagationTestBase is located in WF testsuite, so I think we need to answer duplication question first. 

Anyway we need some sort of high level description what is test actually testing.

File: testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/AbstractSecurityPropagationTestCase.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+    @Deployment(testable = false, name = "WAR", order = 2, managed = true)
+    @TargetsContainer("multinode-client")
+    public static Archive<?> createWarDeploymentUsingRemoteConnections() {
+        return Deployments.createWarDeployment("", false, TestConfig.SERVLET_1);
+    }
+
+    @Test
+    public void testSecurityPropagationSecuredWebClientToRemoteEJB() throws Exception {
+        // unit test (WEB_CREDENTIAL) -> servlet1 (Client Server) (NO Credential) -> ejb1 (Server Server)
```
Comment (rpelisse) [inline testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/AbstractSecurityPropagationTestCase.java]: @mchoma Each test added by @bmaxwell are following by a brief comment describing the test case (as shown above).  Based on this, it seems that maybe this is already by the [testForwardedAuthenticationIdentityDoesNotExistOnIntermediateServer](https://github.com/wildfly/wildfly/blob/master/testsuite/integration/manualmode/src/test/java/org/wildfly/test/manual/elytron/seccontext/ServerChainSecurityContextPropagationTestCase.java#L208) (albeit not using EJB). 

(I'm honestly out of my depth here, I would like to help more, but I'm not sure if I can really assess how the tests relates or not to each other)

Comment (mchoma) [reply testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/security/AbstractSecurityPropagationTestCase.java]: What do you mean by "not using EJB" ?

**Comment by mchoma** [comment]
@bmaxwell I remember these tests get in "last minute" before 7.1.0.GA - looking on dates it was Oct 27, 2017 so that mean after you PR which was created  Sep 5, 2017. So everything what you say is true.

Looking into TeamCity of this PR these tests are currently running - https://ci.wildfly.org/viewLog.html?buildTypeId=WFPR&buildId=117795&tab=testsInfo&pager.currentPage=1

So now questoion is do these tests test something new compared to  tests in org.wildfly.test.manual.elytron.seccontext ?

**Comment by rpelisse** [comment]
> So now questoion is do these tests test something new compared to tests in org.wildfly.test.manual.elytron.seccontext ?

This is is clearly the difficult part to assess. I have a gut feeling that this PR brings a lot of duplication, indeed, with the tests you pointed out, but I don't know how to make the demonstration of it...

**Comment by rpelisse** [comment]
OK, I've tried to look a bit more in depth into the tests provided by @bmaxwell  and compare them to one mentioned by @mchoma. To make it short, I would just say I think they is indeed some overlap but also that the one from @mchoma are a bit more high level. @bmaxwell had to create a series of technical object (EJBRequest, Results...) but the tests from @mchoma seems to have done a better jobs at abstracting those things away (IMHO).

As for the scenario tested, it is difficult for me to assert how much the new tests from @bmaxwell will bring, but if they did bring more coverage, for sure, they would need to be refactor to be integrated into the tests pointed out by @mchoma. 

So to move forward, I'm going to close this pull request and the associated issue as "won't do / already done". @bmaxwell if you feel the current coverage is not enough and want to add some tests, please let me know, I'll try to help you doing so.

**Pull request closed without merging**